### PR TITLE
Time dependent QDM

### PR DESCRIPTION
**Pull request by castelao**
QDM was previously implemented to operate in a full time series, but PresRat requires seasonal CDFs. This was first achieved with a window sliding version of `bias.QuantileDeltaMappingCorrection`, but later I generalized the standard QDM to run on any time periods. That satisfies the case of `n=1` as it was before, as well as `n=12` which is approximately the monthly solution.

The #215 already incorporated these modifications but for the sake of simplicity, this PR should be merged before #215.

### CODE REVIEW

**Comment by castelao** [comment]
@grantbuster @bnb32 , whenever you have a chance, could you review this PR, please? It will make easier for you to review #215.

This PR covers the time dependent solution for QDM. Instead of monthly it is generalized for any time interval. If we use 24 intervals with 30 day time window it would give us still the monthly scale but one estimate every 15 days, which would minimize the jump between on month transitions. Also, with that we avoid having to interpolate when applying and using instead the nearest value. More intervals we add, higher resolution, with an extreme of 365 intervals would lead to one CDF for each day of the year.

File: sup3r/bias/qdm.py
Comment (grantbuster) [inline sup3r/bias/qdm.py:601]: I think this method makes sense but maybe worth adding a small test to make sure... The logic makes sense but it takes a while to understand haha. Should be really lightweight to add a test too.

File: sup3r/bias/bias_transforms.py
Comment (grantbuster) [inline sup3r/bias/bias_transforms.py:530]: Can you update these comments for the new output shape?

**Review by grantbuster — CHANGES_REQUESTED** [review]

File: sup3r/bias/qdm.py
Comment (grantbuster) [reply sup3r/bias/qdm.py:287]: Nevermind i see that it needs to be called in a parallel method

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:254]: That's a good point. It should not fail. This step only gives the window center, so it should be independent of the time interval of `base_data`. We can choose a time window centered anywhere.

When gets to the point to filter the data within a window, `base_ti` is a `pd.DatetimeIndex`, so extracting day of the year should be safe. That would lead to a rounding error since I'm using just `day_of_year` without including the fraction from the hours/minutes. 

I just added a new method `doy_fractional` that we can use or remove it if you're concerned with that rounding error.

About the second part.
If we alternate a fraction of 365 or 366 days, that would result in a slightly variable window. For instance a 1/12 window size would be slightly different for 365 and 366.
 
A more precise solution would be to count continuous time from the start of the time series, divide that running time by 365.25, remove the integer, and use the fraction of an actual year. This would avoid the shift from jumping one year every 4 years. I considered that negligible before. One issue of this approach would be processing `no_leap` type of simulations.

I think that the error should be minimal. The time window is applied everywhere, but any window extending through the change of year would have a few data more to average, slightly biasing the average on that direction.

I suggest keep it as it is, but I'm glad to change if you have a strong opinion about it.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:291]: fixed

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:601]: Certainly. I'm working on that.

**Review by bnb32 — CHANGES_REQUESTED** [review]
Ill let Grant double check the implementation ;) I just have some random comments.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:516]: Modified as requested.

File: sup3r/bias/qdm.py
Comment (grantbuster) [reply sup3r/bias/qdm.py:254]: Can you push? I don't see new commits that you mentioned. 

Okay i think I see your point about how this will actually totally work with hourly data because DOY is being used in the indexing. Maybe note in the _window_center docstring that the output array indexes the day of year? Looks good to me! I'm not worried about fractional days for leap years.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:601]: Done

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:530]: At this point, after applied the `window_idx` in the inputs, these arrays have the same shape and dimensions, so nothing changed here. I'm adding a few more comments anyways.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:287]: I see, not here, but in the documentation. It was updated.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:402]: Sure!

File: sup3r/preprocessing/data_handling/base.py
Comment (castelao) [reply sup3r/preprocessing/data_handling/base.py:1386]: Good catch! Fixed.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:530]: I see, not here, but in the documentation. It was updated.

**Comment by castelao** [comment]
@grantbuster @bnb32 , I'm sorry I forgot to ping you yesterday. I believe that I covered all the requests. Please let me know if it is ready to merge.

**Review by grantbuster — APPROVED** [review]

**Comment by grantbuster** [comment]
@castelao looks good thanks for implementing this!

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:266]: Sure. I'll add that extra info in the documentation in the PresRat branch (after merging with this) so we don't hold this one here any longer.

**Review by bnb32 — APPROVED** [review]

**Pull request merged**
